src/auth/Auth.ts

```diff
@@ -0,0 +1,100 @@
+/* eslint-disable @typescript-eslint/no-unused-vars -- This file contains a base class with a lot
+   of unused paramters. */
+
+import type { ConnectOptions, Credentials } from "@canonical/jujulib";
+
+import type { ConnectionWithFacades } from "juju/types";
+import type { AuthMethod, Credential } from "store/general/types";
+import type { AppDispatch } from "store/store";
+import { logger } from "utils/logger";
+
+/**
+ * Base class for authentication providers. The stubbed lifecycle methods are called throughout the
+ * dashboard. Authentication implementations should extend this class and override the relevant
+ * methods.
+ *
+ * This class should be consumed as a singleton via the `instance` property (eg.
+ * `Auth.instance.logout()`). The singleton is set in the base class's constructor.
+ */
+export class Auth {
+  /**
+   * Singleton instance of the auth provider.
+   */
+  static instance: Auth;
+
+  /**
+   * Redux dispatch method for use within auth implementations.
+   */
+  protected dispatch: AppDispatch;
+
+  /**
+   * Identifier of the auth method.
+   */
+  public name: AuthMethod;
+
+  /**
+   * Create a new instance of the singleton, using the
+   */
+  constructor(dispatch: AppDispatch, name: AuthMethod) {
+    if (Auth.instance) {
+      logger.warn(
+        "Singleton instance already exists, and is being re-constructed. Overriding previous instance with incoming.",
+      );
+    }
+
+    Auth.instance = this;
+    this.dispatch = dispatch;
+    this.name = name;
+  }
+
+  /**
+   * Executed at the conclusion of the dashboard's bootstrap method.
+   */
+  async bootstrap(): Promise<void> {}
+
+  /**
+   * Logout functionality for the authentication provider.
+   */
+  async logout(): Promise<void> {}
+
+  /**
+   * Executed after the controller list is fetched.
+   *
+   * @param conn - Controller connection.
+   */
+  async onControllerListFetched(conn: ConnectionWithFacades): Promise<void> {}
+
+  /**
+   * Executed before a connection is initiated with a controller.
+   *
+   * @returns Whether to continue the connection attempt.
+   */
+  async beforeControllerConnect(
+    controllerData: ControllerData,
+  ): Promise<boolean> {
+    return true;
```

Comment:
Because returning `false` will break the execution in model-poller it might be worth adding a very simple test for this case so that we don't inadvertently break model-poller by changing this.